API changes

max-m opened the pull request:
This issue is meant to track and discuss (proposed) API changes.

Changes so far:

- Changed the return types of environment calls that denote success or failure by returning a `bool` into a proper `Result`

max-m committed:
Replace bool-ey success return types with `Result`s

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -99,21 +148,17 @@ pub unsafe fn set_rotation(callback: retro_environment_t, rotation: Rotation) ->
 /// Boolean value whether or not frontend supports frame duping,
 /// passing NULL to video frame callback.
 #[proc::context(GenericContext)]
-pub unsafe fn can_dupe(callback: retro_environment_t) -> bool {
```
`false` is a valid return value for `RETRO_ENVIRONMENT_GET_CAN_DUPE`, but that property is not preserved here. If `RETRO_ENVIRONMENT_GET_CAN_DUPE` returns false, there's no way to know with this wrapper.

max-m replied:
Whoops. You’re right!

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -194,7 +238,9 @@ pub unsafe fn set_performance_level(callback: retro_environment_t, level: u8) ->
 /// use the new [`get_save_directory()`].
 #[proc::context(GenericContext)]
-pub unsafe fn get_system_directory<'a>(callback: retro_environment_t) -> Option<&'a Path> {
+pub unsafe fn get_system_directory<'a>(
```
`RETRO_ENVIRONMENT_GET_SYSTEM_DIRECTORY` can return `NULL`, but this function provides no way to tell if that was the case. Consider a return value of `Result<Option<&'a Path>, EnvironmentCallError>`.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -209,9 +255,9 @@ pub unsafe fn get_system_directory<'a>(callback: retro_environment_t) -> Option<
 pub unsafe fn set_pixel_format<F: Into<retro_pixel_format>>(
     callback: retro_environment_t,
     format: F,
-) -> bool {
```
`RETRO_ENVIRONMENT_SET_PIXEL_FORMAT` might return `false` if the frontend doesn't support the requested pixel format. Is there something more descriptive available than `EnvironmentCallError::Failure`?

max-m replied:
I suppose we could map this failure case to `EnvironmentCallError::Unsupported(_)` with a message like `the frontend did not accept <pixel format>` or similiar.
But does this provide any meaningful benefit over the general `Failure` case that is used for other environment calls that return `false` as status? :)

JesseTG replied:
Hm

> But does this provide any meaningful benefit over the general `Failure` case that is used for other environment calls that return false as status? :)

If you make the distinction, then the core can try different formats if one fails. If the environment call itself fails, then the core can just not bother.

max-m replied:
I’m still in favor of keeping the error as is. In case of `EnvironmentCallError::Failure` a core is encouraged to try again with a different format it supports.
But I’ll definitely go ahead and change the return type of `set_pixel_format` to return the pixel format on success, this way you can easily chain calls to the function:
```rust

let pixel_format = ctx
    .set_pixel_format(PixelFormat::XRGB8888)
    .or_else(|_| ctx.set_pixel_format(PixelFormat::RGB565))
    .or_else(|_| ctx.set_pixel_format(PixelFormat::XRGB1555))
    .expect(
        "the frondend should support one of these pixel formats: XRGB888, RGB565 or 0RGB1555",
    );
```

Note: `0RGB1555` is the default, so above code should probably we written as
```rust
let pixel_format = ctx
    .set_pixel_format(PixelFormat::XRGB8888)
    .or_else(|_| ctx.set_pixel_format(PixelFormat::RGB565))
    .unwrap_or(PixelFormat::XRGB1555);
```

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -267,9 +312,12 @@ pub unsafe fn set_disk_control_interface(
 /// If HW rendering is used, call either
 /// [`RunContext::draw_hardware_frame`] or [`RunContext::dupe_frame`].
 #[proc::context(LoadGameContext)]
-pub unsafe fn set_hw_render(callback: retro_environment_t, data: retro_hw_render_callback) -> bool {
```
The `struct retro_hw_render_callback` provided to `RETRO_ENVIRONMENT_SET_HW_RENDER` is used for both input (e.g. the context type) and output (all the function pointers). so this function needs to account for that. I'd either modify the provided `retro_hw_render_callback` or return a new one that contains the frontend's modifications. See [here](https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L2735) for details.

max-m replied:
Oh, looks like I forgot that struct when I fixed that for `retro_hw_render_context_negotiation_interface`, which gets stored in the `CoreWrapper`, like the other interface structs that must be kept around.
Interesting that I didn’t notice any problems in the Vulkan example core :face_with_spiral_eyes:

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -282,30 +330,21 @@ pub unsafe fn set_hw_render(callback: retro_environment_t, data: retro_hw_render
-            return None;
-        }
-    };
+pub unsafe fn get_variable<'a>(
```
`RETRO_ENVIRONMENT_GET_VARIABLE` might return a single string if the variable was found, or `NULL` if not. It's also possible to provide a null key, in which case it'll return all variables in a delimited string. See [here](https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L3437-L3443). I think the `data` argument is allowed to be `NULL` too, but I don't know off-hand what it would return.

JesseTG replied:
Perhaps something like this?

```rust
enum RetroVariableResult {
  Value(str),
  CompleteEnvironment(Map<str, str>), // The str's could be slices into the environment string
  NotFound,
}
```

max-m replied:
As far as I can tell RetroArch does not implement this feature. I don’t know whether `libretro.h` is outdated or RetroArch is missing an implementation to return this complete environment string.

Currently I intend to change the code around those environment calls like this:
```rust
// When key is None, sets `retro_variable.key = NULL`, otherwise uses the supplied key
fn get_variable_or_environment<'a>(
    callback: retro_environment_t,
    key: Option<&'a str>,
) -> Result<Option<&'a str>, EnvironmentCallError>

// same as get_variable_or_environment(cb, Some(key))
fn get_variable<'a>(
    callback: retro_environment_t,
    key: &'a str,
) -> Result<Option<&'a str>, EnvironmentCallError>

// calls get_variable_or_environment(cb, None) and parses the value into a key-value map
fn get_all_variables<'a>(
    callback: retro_environment_t,
) -> Result<Option<HashMap<&'a str, &'a str>>, EnvironmentCallError>

// calls RETRO_ENVIRONMENT_SET_VARIABLE with NULL as argument to test for support, as per libretro.h:
// If data is NULL, no changes will be registered and the callback will return true; an implementation may therefore pass NULL in order to test whether the callback is supported.
fn supports_set_variable(callback: retro_environment_t) -> bool
// RetroArch also implements the same behaviour for `RETRO_ENVIRONMENT_GET_VARIABLE`, so there’s also
fn supports_get_variable(callback: retro_environment_t) -> bool
```
That gives `Ok(Some(value))` (frontend returned gave us a `var.value != NULL`), `Ok(None)` (frontend gave us `var.value = NULL`, i.e. variable is unknown) and `Err(_)` in case the callback returns false or some other error happened like failing string conversion.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -360,34 +399,39 @@ pub unsafe fn get_variable<'a>(callback: retro_environment_t, key: &'a str) -> O
 /// Result is set to [`true`] if some variables are updated by
 /// frontend since last call to [`get_variable`].
 #[proc::context(GenericContext)]
-pub unsafe fn get_variable_update(callback: retro_environment_t) -> bool {
```
It's not an error if `RETRO_ENVIRONMENT_GET_VARIABLE_UPDATE` returns `false`; it just means that no variables have changed. (Not to be confused with the `retro_environment_t` itself returning `false`.) See [here](https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L656-L661) for details.

max-m replied:
Right, this should have been `Result<bool, EnvironmentCallError>`.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -398,7 +442,9 @@ pub unsafe fn set_support_no_game(callback: retro_environment_t, value: bool) ->
 /// Mostly useful in cooperation with [`set_support_no_game`] as assets can
 /// be loaded without ugly hacks.
 #[proc::context(GenericContext)]
-pub unsafe fn get_libretro_path<'a>(callback: retro_environment_t) -> Option<&'a Path> {
```
`RETRO_ENVIRONMENT_GET_LIBRETRO_PATH` can return `NULL` if libretro itself was loaded statically, or if it's built into the frontend. I think the return type should be `Result<Option<&'a Path>, EnvironmentCallError>`.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -99,21 +148,17 @@ pub unsafe fn set_rotation(callback: retro_environment_t, rotation: Rotation) ->
 )]
 #[proc::context(GenericContext)]
-pub unsafe fn get_overscan(callback: retro_environment_t) -> bool {
+pub unsafe fn get_overscan(callback: retro_environment_t) -> Result<(), EnvironmentCallError> {
```
`RETRO_ENVIRONMENT_GET_OVERSCAN` can return `false`, it's not an error. (However, the `retro_environment_t` itself returning `false` is still an error.)

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -463,9 +509,9 @@ pub unsafe fn set_audio_callback(
 pub unsafe fn get_rumble_interface(
     callback: retro_environment_t,
-) -> Option<retro_rumble_interface> {
+) -> Result<retro_rumble_interface, EnvironmentCallError> {
```
`RETRO_ENVIRONMENT_GET_RUMBLE_INTERFACE` might return null if rumble functionality isn't available. I would argue that that isn't an error.

max-m replied:
```c
case RETRO_ENVIRONMENT_GET_RUMBLE_INTERFACE:
{
   struct retro_rumble_interface *iface =
      (struct retro_rumble_interface*)data;

   RARCH_LOG("[Environ]: GET_RUMBLE_INTERFACE.\n");
   iface->set_rumble_state = input_set_rumble_state;
   break;
}
```
this is one of the environment calls that asks you to supply the struct and the frontend fills it out for you.
So I’d argue that this implementation is fine. We create the `retro_rumble_interface` struct, pass the pointer to the frontend which either implements the callback, sets `set_rumble_state` and returns true or returns false when the feature is unsupported. The former case is `Ok(retro_rumble_interface)`, the latter `Err(EnvironmentCallError::Failure)`.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -577,7 +625,9 @@ pub unsafe fn get_location_callback(
 /// If so, no such directory is defined,
 /// and it's up to the implementation to find a suitable directory.
 #[proc::context(GenericContext)]
-pub unsafe fn get_core_assets_directory<'a>(callback: retro_environment_t) -> Option<&'a Path> {
```
`RETRO_ENVIRONMENT_GET_CORE_ASSETS_DIRECTORY` can return `NULL`, so I'd make the return value a `Result<Option<&'a Path>, EnvironmentCallError>`.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -1000,31 +1068,33 @@ pub fn get_led_interface(callback: retro_environment_t) -> Option<retro_led_inte
 #[proc::context(GenericContext)]
 #[proc::unstable(feature = "env-commands")]
-pub unsafe fn get_audio_video_enable(callback: retro_environment_t) -> AudioVideoEnable {
+pub unsafe fn get_audio_video_enable(
```
~~The `AudioVideoEnable` flags are used for both input and output.~~

No, wait, I'm wrong. Output only.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -1869,13 +1974,11 @@ pub unsafe fn set_content_info_override(
 ///   [`Core::on_load_game_special`]
 #[proc::context(LoadGameContext)]
 #[proc::context(LoadGameSpecialContext)]
-pub unsafe fn get_game_info_ext(callback: retro_environment_t) -> Option<retro_game_info_ext> {
```
This is supposed to return an array. https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L1477-L1510 Or I guess in this case, a slice; the frontend maintains ownership of the underlying `retro_game_info_ext` array, so it'll be fine.

JesseTG reviewed (commented):
~~I would argue that most (if not all) of the wrappers in `environment.rs` could be considered safe, as any `unsafe` considerations like pointer casts appear to be handled internally.~~

Hm, actually, maybe not. I'm not sure yet.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -1828,9 +1933,9 @@ pub unsafe fn set_fastforwarding_override(
-) -> bool {
+) -> Result<(), EnvironmentCallError> {
     // const struct retro_system_content_info_override *
-    set(callback, RETRO_ENVIRONMENT_SET_CONTENT_INFO_OVERRIDE, value).unwrap_or(false)
```
`RETRO_ENVIRONMENT_SET_CONTENT_INFO_OVERRIDE` is supposed to accept an array (or I guess a slice, in this case).

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -1680,14 +1784,13 @@ pub unsafe fn get_input_max_users(callback: retro_environment_t) -> (u32, bool)
     callback: retro_environment_t,
     data: retro_audio_buffer_status_callback,
-) -> bool {
+) -> Result<(), EnvironmentCallError> {
```
`NULL` (or I guess in this case, `None`) is an acceptable input parameter for `RETRO_ENVIRONMENT_SET_AUDIO_BUFFER_STATUS_CALLBACK`; it disables the callback. https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L1358-L1365

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -474,14 +520,17 @@ pub unsafe fn get_rumble_interface(
-    }
+    let caps = get::<u64>(callback, RETRO_ENVIRONMENT_GET_INPUT_DEVICE_CAPABILITIES)?;
+
+    // I’m not entirely sure why this call returns a 64 bit value when the `RETRO_DEVICE_MASK`
```
I think the return value is 64-bit to allow extra types of input devices to be added without breaking ABI compatibility.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -595,7 +645,9 @@ pub unsafe fn get_core_assets_directory<'a>(callback: retro_environment_t) -> Op
 /// files. Cores that need to be backwards-compatible can still check
 /// [`get_system_directory`].
 #[proc::context(GenericContext)]
-pub unsafe fn get_save_directory<'a>(callback: retro_environment_t) -> Option<&'a Path> {
```
`RETRO_ENVIRONMENT_GET_SAVE_DIRECTORY` can return `NULL` to indicate that no save data directory is available. This is expected behavior, so I'd make the return type something like `Result<Option<&'a Path>, EnvironmentCallError>`.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -788,34 +841,32 @@ pub unsafe fn set_game_geometry(
 /// If this environment callback is used by a core that requires a valid username,
 /// a default username should be specified by the core.
 #[proc::context(GenericContext)]
-pub unsafe fn get_username<'a>(callback: retro_environment_t) -> Option<&'a str> {
```
`NULL` is a valid return value for `RETRO_ENVIRONMENT_GET_USERNAME`, so this should return something like a `Result<Option<&'a str>, EnvironmentCallError>`.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -939,14 +1005,13 @@ pub unsafe fn set_hw_render_context_negotiation_interface(
     callback: retro_environment_t,
     quirks: SerializationQuirks,
-) -> bool {
+) -> Result<(), EnvironmentCallError> {
```
`RETRO_ENVIRONMENT_SET_SERIALIZATION_QUIRKS` returns a modified set of quirks flags, because the frontend is expected to clear any bits that it doesn't recognize.

max-m replied:
`SerializationQuirks` should also be based on `u64`, I’ll fix this as well.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -959,14 +1024,15 @@ pub unsafe fn set_serialization_quirks(
 /// being used.
 #[proc::context(GenericContext)]
 #[proc::unstable(feature = "env-commands")]
-pub unsafe fn set_hw_shared_context(callback: retro_environment_t) -> bool {
```
`RETRO_ENVIRONMENT_SET_HW_SHARED_CONTEXT` can return `false` to indicate that shared hardware contexts aren't supported. I'd argue that asking this question and getting a "no" isn't an error.

max-m replied:
Since this function doesn't take any data and just returns `true` or `false` for the callback, we end up with
`if ctx.set_hw_shared_context().is_ok() { ... }` vs. `if let Ok(true) = ctx.set_hw_shared_context() { ... }`
of those two cases I prefer the first, both give the same knowledge: Either the call succeeded or it failed.
The former knows two error cases: The environment call function pointer was a null pointer or the callback returned `false` (which is the general error case for the environment calls).
The latter replaces the second error case with `Ok(false)` and requires more work to check if the call succeeded.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -979,17 +1045,19 @@ pub unsafe fn set_hw_shared_context(callback: retro_environment_t) -> bool {
     callback: retro_environment_t,
     data: retro_vfs_interface_info,
-) -> Option<retro_vfs_interface_info> {
+) -> Result<retro_vfs_interface_info, EnvironmentCallError> {
```
The only required input to `RETRO_ENVIRONMENT_GET_VFS_INTERFACE` is the version number, so I'd argue that you could simplify the use of this function by only accepting a version number.

max-m replied:
`SetEnvironmentContext::enable_vfs_interface(min_version)` provides a wrapper around this function that also stores the struct in the `CoreWrapper` for you and enables functions like `GenericContext::vfs_get_path()` to work.
So I don’t think that changing the API is necessary here.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -1000,31 +1068,33 @@ pub fn get_led_interface(callback: retro_environment_t) -> Option<retro_led_inte
 /// Boolean value that indicates whether or not the frontend is in fastforwarding mode.
 #[proc::context(GenericContext)]
 #[proc::unstable(feature = "env-commands")]
-pub unsafe fn get_fastforwarding(callback: retro_environment_t) -> bool {
```
`RETRO_ENVIRONMENT_GET_FASTFORWARDING` returning `false` is not an error or an indicator of a lack of support. It just means that the player isn't holding the "fast-forward" button. So I'd return a `Result<bool, EnvironmentCallError>`.

max-m committed:
`can_dupe()` should return a `bool`

max-m committed:
Change `get_system_directory()` to return an Option on success

From `libretro.h`:
> The returned value can be NULL.
> If so, no such directory is defined,
> and it's up to the implementation to find a suitable directory.

JesseTG commented on rust-libretro/src/environment.rs:
```diff
@@ -1921,18 +2023,23 @@ pub unsafe fn set_core_options_update_display_callback(
 /// to test whether the callback is supported.
 #[proc::context(GenericContext)]
-pub unsafe fn set_variable(callback: retro_environment_t, value: retro_variable) -> bool {
+pub unsafe fn set_variable(
```
`RETRO_ENVIRONMENT_SET_VARIABLE` accepts an input of `NULL` to test for support.

JesseTG reviewed (changes requested):
Thanks for your patience, and for putting the work in to this crate. These are all of my thoughts on the API changes.

max-m committed:
Make sure we keep `retro_hw_render_callback` after calling `RETRO_ENVIRONMENT_SET_HW_RENDER`

max-m committed:
Fix `wrapper.supports_bitmasks` and `wrapper.core.on_options_changed()`

max-m committed:
Make `GenericContext::get_variable()` return None when the option is unknown

max-m committed:
Return `Option<Path>` for nullable paths like `RETRO_ENVIRONMENT_GET_LIBRETRO_PATH`

max-m committed:
Make `GenericContex::get_overscan()` return a `Result<bool, E:wq>`

max-m committed:
Fix comment for environment::get_unchecked<T>()

max-m committed:
Fix example core "test"

max-m committed:
Fix `get_game_info_ext()` and add `get_game_info_ext_array()`

max-m committed:
Remove comment in `get_input_device_capabilities()` and add optional stricter validation

max-m committed:
Fix return type of `environment::get_fastforwarding()`

max-m committed:
`environment::set_serialization_quirks()` is allowed to update the provided flags

max-m committed:
`environment::get_username()` is allowed to return no name

max-m committed:
Make `environment::set_audio_buffer_status_callback()` accept a nullable value

max-m committed:
Fix some comments in environment.rs

max-m committed:
`environment::set_content_info_override()` expects a null terminated array

max-m committed:
Remove user-visible RetroArch workaround from `Core::on_set_environment()`

max-m committed:
Fix weak rumble effect in example core "test"

max-m committed:
Return the pixel format in `environment::set_pixel_format()` on success

This allows chaining this function in case a core supports multiple pixel formats:

```rust
let pixel_format = ctx
    .set_pixel_format(PixelFormat::XRGB8888)
    .or_else(|_| ctx.set_pixel_format(PixelFormat::RGB565))
    .unwrap_or(PixelFormat::XRGB1555);
```

max-m committed:
Fix some Clippy lints

max-m committed:
Remove useless `is_null()` check

max-m committed:
[CI] Run either the `push` or `pull_request` action but not both

max-m committed:
[CI] Allow workflows to be triggered manually

JesseTG reviewed (approved):
It all looks good to me. :+1: No further opinions; if I find anything else, it'll be while using it.
